### ISSUE TITLE
Enable kustomize to effectively combine multiple kustomizations

### ISSUE THREAD

**Issue by yanniszark** [kubernetes-sigs/kustomize#3710]
**Is your feature request related to a problem? Please describe.**

Kubeflow is a big consumer and supporter of kustomize. I am opening this issue to address problems I have bumped into, while trying to finalize the release of Kubeflow 1.3. Note that I am the release manager for Kubeflow 1.3 and also the lead of Kubeflow's Manifests Working Group.

In Kubeflow we used `kfctl` in the past to combine multiple kustomizations together. `kfctl` is a custom Kubeflow tool we used to cover missing features of kustomize, but we have deprecated it for 1.3. We now want to use upstream kustomize to combine multiple kustomizations, as this is a feature that kustomize already has. However, a feature called `vars` prevents core kustomize from combining multiple kustomizations.

We have worked around this issue by extending kustomize's code to enable importing kustomizations as pure yaml. In other words, importing the result of `kustomize build`. This enables us to combine kustomizations and we would really like for it to become part of core kustomize, to both help with our 1.3 release, but also use it heavily in the next releases as well.

**Why current kustomize doesn't work:**

Kustomize can already combine kustomizations via the `resources` field:
```yaml
resources
- /path/to/kustomization1
- /path/to/kustomization2
```

Kustomize also has a feature called `vars`, which we use extensively in Kubeflow's codebase. The main use of `vars` is to declare cross-references (e.g., the namespace for the service fqdn in the VirtualService's spec is the same as the VirtualService's namespace).

However, `vars` break the encapsulation of the kustomization and must be unique on a global level. Thus, a kustomization can't import two kustomizations that declare the same var.

This would crash:
```yaml
resources
- /path/to/kustomization1  # declares var "example"
- /path/to/kustomization2  # declares var "example"
```

**Describe the solution you'd like**

A field to import inflated kustomizations. In other words, import the result of running `kustomize build` on a given set of paths.
Here is what it would look like:
```yaml
kustomizationPaths:
- <kustomization_1>
- <kustomization_2>
- glob:<folder_with_kustomizations>/*
```

**Comment by KnVerey** [comment]
Am I understanding correctly that the `resources` approach works fine if `vars` are not used? There are plans to deprecate the `vars` field (https://github.com/kubernetes-sigs/kustomize/issues/2052) and replace it with a transformer (https://github.com/kubernetes-sigs/kustomize/issues/3492). Assuming the deprecation proceeds, I imagine we shouldn't add a new field to work around `vars` issues. 

cc @natasha41575 since you're working on the transformer.

**Comment by aodinokov** [comment]
@yanniszark  JFYI, Alternatively this may be done via external plugins. (This info may be helpful if this feature isn't accepted)
e.g.
Kustomize allows [krm-functions](https://googlecontainertools.github.io/kpt/guides/producer/functions/) to be used as plugins. - this is pretty convenient since it doesn't require any kustomize configuration. everything will be done by docker image, that is pulled automatically.
 
[Here](https://github.com/kubernetes-sigs/kustomize/blob/master/api/krusty/fnplugin_test.go#L111) is the example in the tests how you can write the config for krm function. Probably it's possible to write the needed plugin/function.
Moreover at some point there was a [kustomize_build krm-function](https://github.com/GoogleContainerTools/kpt-functions-catalog/blob/78889147641e942c6307ba655088a0258ab5dc01/functions/ts/src/kustomize_build.ts) ([link to the docker image](https://quay.io/repository/aodinokov/kustomize_build_dind)). [Here](https://github.com/GoogleContainerTools/kpt/issues/879) is the discussion how to use that. Actually it worked even though it was hard to understand how to run it :) The only difference from the discussion I mentioned will be that you will put fn config into yaml that you'll need to use as generator config. and you'll need to pass needed params not to kpt fn run, but to kustomize build --allow-alpha-plugins...

Basically that allows to call kustomize build (as a krm function) that will `inflate` resources from another kustomize build without issues with vars.

**Comment by monopole** [comment]
@knverey is right
the `vars` is not supported, and removing it is coming up the stack in importance.  It goes away as soon as we have a migration plan in place to the replacement transformer.


Also, if an entry in `resources` is a path to a directory (or a path to a directory in a github repo) it is recursively expanded as a `kustomize build`, before being considered as part of the existing kustomization.

So to incorporate a bunch of yaml in some directory, just drop a kustomization file into that directory.  The kustomize program has `create` and `add` commands (the latter accepts globbing) to help with this.

**Comment by yanniszark** [comment]
Thanks for your feedback Jeff, we unfortunately caught this late in our cycle and we need it for kubeflow 1.3. We're thinking of shipping this as a temporary fork of kustomize and then immediately rebasing on this as a follow on point release or next major release, what do you think? Or is there some other way we can workaround that might make this simpler?

**Comment by komapa** [comment]
@yanniszark sorry to  resurrect this old issue but what is the current way you are solving this exact problem in KubeFlow with kustomize? Thank you!
